pkgs/by-name/ap/apple-sdk/scripts/lock-sdk-deps.sh

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p coreutils curl git gnutar jq moreutils nix
+
+set -eu -o pipefail
+
+if [ ! -v 2 ]; then
+    echo "usage: lock-sdk-deps.sh <SDK version> <Packages>" >&2
+    echo " <SDK version>   Decimal-separated version number." >&2
+    echo "                 Must correspond to a tag in https://github.com/apple-oss-distributions/distribution-macOS" >&2
+    echo " <Packages>      List of packages from the distributions-macOS repository." >&2
+    echo "                 Packages not in the repository at the tag for <SDK version> will be ignored."
+    exit 1
+fi
+
+pkgdir=$(dirname "$(dirname "$(realpath "$0")")")
+
+lockfile=$pkgdir/metadata/apple-oss-lockfile.json
+if [ -e "$lockfile" ]; then
+    touch "$lockfile"
+fi
+
+workdir=$(mktemp -d)
+trap 'rm -rf -- "$workdir"' EXIT
+
+sdkVersion=$1; shift
+tag="macos-${sdkVersion//.}"
+
+declare -a packages=("$@")
+
+echo "Locking versions for macOS $sdkVersion using tag '$tag'..."
+
+pushd "$workdir" > /dev/null
+
+git clone --branch "$tag" https://github.com/apple-oss-distributions/distribution-macOS.git &> /dev/null
+cd distribution-macOS
+
+for package in "${packages[@]}"; do
+    # If the tag exists in `release.json`, use that as an optimization to avoid downloading unnecessarily from Github.
+    packageTag=$(jq -r --arg package "$package" '.projects[] | select(.project == $package) | .tag' release.json)
+    packageCommit=$(git ls-tree -d HEAD "$package" | awk '{print $3}')
+
+    if [ ! -d "$package" ]; then
+        packageCommit=HEAD
+    fi
+
+    # However, sometimes it doesn’t exist. In that case, fall back to cloning the repo and check manually
+    # which tag corresponds to the commit from the submodule.
+    if [ -z "$packageTag" ]; then
+        git clone --no-checkout "https://github.com/apple-oss-distributions/$package.git" ../source &> /dev/null
+        pushd ../source > /dev/null
+        packageTag=$(git tag --points-at "$packageCommit")
+        popd > /dev/null
+        rm -rf ../source
+    fi
+
+    packageVersion=${packageTag##"$package"-}
+
+    curl -OL "https://github.com/apple-oss-distributions/$package/archive/$packageTag.tar.gz" &> /dev/null
+    tar axf "$packageTag.tar.gz"
```

Comment:
I’m not following what the shortcut would be.

Reply:
Checking out the commit.

Reply:
The repo isn’t necessarily cloned already. That’s only done if Apple hasn’t updated the `release.json` file, and the script has to get the dependency out of `.gitmodules`. Having to do that is pretty unusual, so it’s not optimized to avoid downloading the tarball.

However, if I did do that (checking out the commit in that case), I’d also have to scrub the repo to make sure it didn’t have any files it shouldn’t (such as `.git`). It adds complexity for a rare case, which I don’t think is worth it.